src/editors/playgroundResultProvider.ts

```diff
@@ -1,44 +1,145 @@
+            this._playgroundResult.content[index][item] = data[item];
+          }
+        });
+      }
+    } else if (
+      this._playgroundResult.content !== null &&
+      typeof this._playgroundResult.content === 'object'
+    ) {
+      const content = { ...this._playgroundResult.content };
+
+      Object.keys(content).forEach((item) => {
+        if (this._playgroundResult && this._playgroundResult.content) {
+          this._playgroundResult.content[item] = data[item];
+        }
+      });
+    }
+
+    if (this._uri) {
+      this.onDidChangeEmitter.fire(this._uri);
+    }
+  }
+
+  provideTextDocumentContent(uri: vscode.Uri): Promise<string> {
+    this._uri = uri;
+
     return new Promise((resolve) => {
-      this._statusView.showMessage('Getting results...');
-
-      if (
-        typeof this._playgroundController.playgroundResult?.content ===
-          'object' ||
-        this._playgroundController.playgroundResult?.type !== 'string'
-      ) {
-        return resolve(
-          JSON.stringify(
-            this._playgroundController.playgroundResult?.content,
-            null,
-            2
-          )
-        );
+      const namespace: string | null = this._playgroundResult.namespace;
+      const type: string | null = this._playgroundResult.type;
+      const content: any = this._playgroundResult.content;
+
+      if (type === 'undefined') {
+        return resolve('undefined');
+      }
+
+      if (type === 'string') {
+        return resolve(this._playgroundResult.content);
       }
 
-      return resolve(this._playgroundController.playgroundResult?.content);
+      const codeLensesInfo: DocCodeLensesInfo = [];
+
+      // Show code lenses only for the list of documents returned by `find()`
+      // or the single document returned by `findOne()`.
+      if (type === 'Cursor' && Array.isArray(content)) {
+        // When the playground result is the collection,
+        // show the first code lense after [{.
+        let line = 2;
+
+        content.forEach((item) => {
+          // We need _id and namespace for code lenses
+          // to be able to save the editable document.
+          if (item._id && namespace) {
+            codeLensesInfo.push({ line, documentId: item._id, namespace });
+            // To calculate the position of the next open curly bracket,
+            // we stringify the object and use a regular expression
+            // so we can count the number of lines.
+            line += JSON.stringify(item, null, 2).split(/\r\n|\r|\n/).length;
+          }
+        });
+      } else if (type === 'Document' && content._id && namespace) {
+        // When the playground result is the single document,
+        // show the single code lense after {.
+        codeLensesInfo.push({ line: 1, documentId: content._id, namespace });
+      }
+
+      this._editDocumentCodeLensProvider?.refresh(codeLensesInfo);
+
+      return resolve(JSON.stringify(content, null, 2));
```

Comment:
I might be missing something - maybe this function be synchronous?

Reply:
Nice observation! Actually, I took as an example CollectionViewProvider and `provideTextDocumentContent` there is a Promise. But according to documentation can have different responses types: `T | undefined | null | Thenable<T | undefined | null>`. Refactored to `provideTextDocumentContent(): string` and `return JSON.stringify(content, null, 2);`.